yt_dlp/extractor/instagram.py

```diff
@@ -58,14 +58,18 @@ def _perform_login(self, username, password):
             self._LOGIN_URL, None, note='Downloading login webpage', errnote='Failed to download login webpage')
 
         shared_data = self._parse_json(self._search_regex(
-            r'window\._sharedData\s*=\s*({.+?});', login_webpage, 'shared data', default='{}'), None)
+            r'\["XIGSharedData",\s*\[\],\s*({.+?}),\d+\]', login_webpage, 'shared data', default='{}'), None)
+        if shared_data.get('native', {}).get('config', {}).get('csrf_token'):
+            shared_data = shared_data['native']
+        else:
+            shared_data = json.loads(shared_data.get('raw', '{}'))
```

Comment:
Use traverse_obj. Read "optional and mandatory fields" in contributing guide

Reply:
imo we don't want anything about the first pass of the JSON (`native`) to be fatal, but if we can't find the required header data on the 2nd pass (`raw`) we want it to be fatal to prevent sending a bad login attempt
```suggestion
            [r'window\._sharedData\s*=\s*({.+?});', r'\["XIGSharedData",\s*\[\],\s*({.+?}),\d+\]'],
            login_webpage, 'shared data', default='{}'), None)

        csrf_token = traverse_obj(shared_data, ('native', 'config', 'csrf_token', {str}))
        rollout_hash = traverse_obj(shared_data, ('native', 'rollout_hash', {str}))
        if not csrf_token or not rollout_hash:
            shared_data = self._parse_json(shared_data['raw'], None)
```
to be applied w/ my other suggestion below